base/number.jl

```diff
@@ -63,4 +63,16 @@ zero{T<:Number}(::Type{T}) = convert(T,0)
 one(x::Number)  = oftype(x,1)
 one{T<:Number}(::Type{T}) = convert(T,1)
 
+promote_op{R,S<:Number}(::Type{R}, ::Type{S}) = (@_pure_meta; R) # to fix ambiguities
+function promote_op{T<:Number}(op, ::Type{T})
+    S = typeof(op(one(T)))
+    # preserve the most general (abstract) type when possible
+    return isleaftype(T) ? S : typejoin(S, T)
+end
+function promote_op{R<:Number,S<:Number}(op, ::Type{R}, ::Type{S})
```

Comment:
I now dispatching in type is more Julian. The problem here, if I understand Jeff right, is that any method that ignores op, that is `promote_op(op, ...)` should preferably return `Any`, and that we should have as fewer such methods as possible (but also the as fewer as possible of the ones that take op into account).

The methods that have `Number` types as arguments (other than op) are a special case since we used in other places beside broadcast. The methods with `Nullable` types are probably also important on its own.

The comparison methods are failing because there is no `isless` for `Complex` and other numerical types.